libavfi/dnn: add LibTorch as one of DNN backend

Semmer2 opened the pull request:
PyTorch is an open source machine learning framework that accelerates
the path from research prototyping to production deployment. Official
websit: https://pytorch.org/. We call the C++ library of PyTorch as
LibTorch, the same below.

To build FFmpeg with LibTorch, please take following steps as reference:
1. download LibTorch C++ library in https://pytorch.org/get-started/locally/,
please select C++/Java for language, other options as your need.
2. unzip the file to your own dir, with command
unzip libtorch-shared-with-deps-latest.zip -d your_dir
3. export your_dir/libtorch/include and
your_dir/libtorch/include/torch/csrc/api/include to $PATH
export your_dir/libtorch/lib/ to $LD_LIBRARY_PATH
4. config FFmpeg with ../configure --enable-libtorch --extra-cflag=-I/your_dir/libtorch/include --extra-cflag=-I/your_dir/libtorch/include/torch/csrc/api/include --extra-ldflags=-L/your_dir/libtorch/lib/
5. make

To run FFmpeg DNN inference with LibTorch backend:
./ffmpeg -i input.jpg -vf dnn_processing=dnn_backend=torch:model=LibTorch_model.pt -y output.jpg
The LibTorch_model.pt can be generated by Python with torch.jit.trace() api.

Semmer2 committed:
libavfi/dnn: add LibTorch as one of DNN backend

PyTorch is an open source machine learning framework that accelerates
the path from research prototyping to production deployment. Official
websit: https://pytorch.org/. We call the C++ library of PyTorch as
LibTorch, the same below.

To build FFmpeg with LibTorch, please take following steps as reference:
1. download LibTorch C++ library in https://pytorch.org/get-started/locally/,
please select C++/Java for language, other options as your need.
2. unzip the file to your own dir, with command
unzip libtorch-shared-with-deps-latest.zip -d your_dir
3. export your_dir/libtorch/include and
your_dir/libtorch/include/torch/csrc/api/include to $PATH
export your_dir/libtorch/lib/ to $LD_LIBRARY_PATH
4. config FFmpeg with ../configure --enable-libtorch --extra-cflag=-I/your_dir/libtorch/include --extra-cflag=-I/your_dir/libtorch/include/torch/csrc/api/include --extra-ldflags=-L/your_dir/libtorch/lib/
5. make

To run FFmpeg DNN inference with LibTorch backend:
./ffmpeg -i input.jpg -vf dnn_processing=dnn_backend=torch:model=LibTorch_model.pt -y output.jpg
The LibTorch_model.pt can be generated by Python with torch.jit.trace() api.

Semmer2 commented on libavfilter/dnn/dnn_backend_torch.cpp:
```diff
@@ -0,0 +1,535 @@
+#define HEIGHT 360
+#define WIDTH 639
+#define UP_SCALE_TIMES 4
+
```
Fixed input size and upscale times, still trying to get input tensor for now.

Semmer2 commented on libavfilter/dnn/dnn_backend_torch.cpp:
```diff
@@ -0,0 +1,535 @@
+    if (request->input_tensor)
+        delete(request->input_tensor);
+    if (request->output)
+        delete(request->output);
```
Segmentation fault may occur when delete output tensor.

xuguangxin commented on configure:
```diff
@@ -5304,9 +5306,9 @@ if test "$?" != 0; then
 fi
 
-add_cppflags -D_ISOC99_SOURCE
+add_cppflags -D_GLIBCXX_USE_CXX11_ABI=0 -D_ISOC99_SOURCE
```
This will not accept by upstream. Why do we need to change this?

xuguangxin commented on configure:
```diff
@@ -6489,6 +6491,7 @@ enabled libopus           && {
     }
 }
 enabled libpulse          && require_pkg_config libpulse libpulse pulse/pulseaudio.h pa_context_new
+enabled libtorch          && require_cpp libtorch torch/torch.h "torch::Tensor" -ltorch -lc10 -ltorch_cpu -lstdc++ -lpthread
```
libtorch has package config file. better use it. https://pytorch.org/cppdocs/installing.html

xuguangxin commented on libavfilter/dnn/dnn_backend_torch.cpp:
```diff
@@ -0,0 +1,535 @@
+        }
+        ff_queue_destroy(th_model->lltask_queue);
+
+        while (ff_queue_size(th_model->task_queue) != 0) {
```
need ff_queue_destroy?

xuguangxin commented on libavfilter/dnn_filter_common.c:
```diff
@@ -53,19 +53,31 @@ static char **separate_output_names(const char *expr, const char *val_sep, int *
-    if (!ctx->model_outputnames) {
-        av_log(filter_ctx, AV_LOG_ERROR, "could not parse model output names\n");
-        return AVERROR(EINVAL);
+    if (backend == DNN_TH) {
```
why tf and openvino not need this?

xuguangxin commented on libavfilter/dnn_filter_common.c:
```diff
@@ -113,8 +125,11 @@ DNNReturnType ff_dnn_get_input(DnnContext *ctx, DNNData *input)
 {
-    return ctx->model->get_output(ctx->model->model, ctx->model_inputname, input_width, input_height,
-                                    (const char *)ctx->model_outputnames[0], output_width, output_height);
+    DNNBackendType backend = ctx->backend_type;
```
why pytroch is so speical ?

Semmer2 replied:
> why pytroch is so speical ?

Since pytorch C++ api do not need input/output names to do the inference.

xuguangxin commented on libavfilter/dnn/dnn_backend_torch.cpp:
```diff
@@ -0,0 +1,535 @@
+        tensor = NULL;
+        return;
+    }
+    //convert between packed and planner RGB24
```
seems we have simliar function in swscale, gbr24ptopacked24. 
could we use swscale do this?

Semmer2 replied:
> seems we have simliar function in swscale, gbr24ptopacked24. could we use swscale do this?

Will add code to support function like gbr24ptopacked24

Semmer2 committed:
libavutil: Add RGBP pixel formats

RGBP means planar RGB format.

FFmpeg supports conversion between GBRP and other RGB formats, but does not support
conversion between RGBP and other RGB formats. So add RGBP pixel format for
following conversion supporting.

Semmer2 committed:
libswscale: Add support for conversion between RGBP and RGB24

Add support for conversion between RGBP(RGB planar) and RGB24(RGB packed)

Semmer2 committed:
libavfi/dnn: refine structure of DNNData

Change DNNData member DNNColorOrder order to enum AVPixelFormat format.

sramakintel commented:
Is this shared library or static? Seems to be shared as the package contains libtorch.so, any ways to build libtorch.a and integrate with FFmpeg?
